src/twisted/web/server.py

```diff
@@ -878,3 +884,110 @@ def acceptableProtocols(self):
             baseProtocols.insert(0, b"h2")
 
         return baseProtocols
+
+
+class _HTTPSRedirectProtocolWrapper(ProtocolWrapper):
+    """
+    Wraps the TLS protocol and redirect to HTTPS if an HTTP request is
+    detected after a connection is made.
+    """
+
+    def dataReceived(self, data):
+        """
+        Only called when initial connection data is received over SSL.
+
+        After that, the dataReceived method of the wrapped protocol is called.
+
+        It checks if an HTTP request is made over HTTPS port, instead of an
+        SSL handshake.
+
+        @param data: The data received from the remote HTTP client peer.
+        @type data: L{bytes}.
+        """
+
+        def get_http_host(data):
+            """
+            Return the value of the Host header found inside `data` HTTP
+            raw request.
+
+            @param data: The first data chunk received from the remote HTTP client peer.
+            @type data: L{bytes}.
+
+            @return: The value of the Host header. It returns the empty string when the value could not be extracted.
+            @rtype: L{bytes}
+            """
+            parts = data.split(b"\r\n")
+            if len(parts) == 1:
+                return b""
+
+            for part in parts:
+                if not part.lower().startswith(b"host"):
+                    continue
```

Comment:
The `Host` header isn't necessary *guaranteed* to show up in the first packet, so this is a bit of a gross heuristic.  I wonder if this might not be better served by a precompiled binary regex, rather than splitting everything up into lines, since we only want to process *one* header here?

Reply:
True.

The use case here is an HTTP request made by a web browser.
AFAIK all web browser will send the host header.

The use case for this patch is confused users that forget to use `HTTPS://` when accessing a web server.

If you don't have a host, it probably means  that you have a custom HTTP client and in that case you don't want a redirect.